Update Ansible processing to the final format

lzap opened the pull request:
The new format has now "results" element with "result" and "task" subelements. This is ready for the future (Ansible may add more fields later on). In the root of the result hash, processor stores new fields which can be used in foreman.

Also add one new keyword.

lzap committed:
Update Ansible processing to the final format

ofedoren commented on lib/smart_proxy_host_reports/ansible_processor.rb:
```diff
@@ -82,46 +74,56 @@ def spool_report
+  def process_level(result)
+    if result["failed"]
+      result["level"] = "err"
+    elsif result["changed"]
```
```suggestion
    elsif result["result"]["changed"]
```
The `changed` key is not on the same level as `failed` key. Have no idea why, but...

![ScreenShot-1633553870196](https://user-images.githubusercontent.com/32508194/136282091-85ccef21-0ebb-426f-81bd-44ef85d473a5.png)

lzap replied:
Yeah you are right, now it works as expected.

lzap replied:
JFYI "result" and "task" are coming from Ansible unchanged, some keys in the root are created in the callback.py. And some others will be added by this plugin.

ofedoren replied:
You probably forgot to update this line in the same way as here: https://github.com/theforeman/smart_proxy_host_reports/pull/5/files#diff-73551ae351ce5f8b8df387a12af7654fa8789cfed325641f8cbf48311e49b028R80

lzap replied:
Yeeah right.

ofedoren commented on lib/smart_proxy_host_reports/ansible_processor.rb:
```diff
@@ -82,46 +74,56 @@ def spool_report
-      module_args = log["invocation"]["module_args"]
-      "Copy #{module_args["_original_basename"]} to #{log["dest"]}"
+    case task_tree["action"]
+    when "ansible.builtin.package"
```
```suggestion
    when "package"
```

Let's stick with `foreman_ansible` approach. `ansible.builtin.module` appears if use ansible via command line and use built-in modules. For custom roles/tasks it's missing.

lzap replied:
I am not sure what you mean, I just tested this with:

```
ANSIBLE_LOAD_CALLBACK_PLUGINS=1 sudo ansible localhost -vvv -m ansible.builtin.package -a "name=vim state=present"

{
  "format": "ansible",
  "id": "73985b36-7a1e-4175-8acc-4b3630113c7a",
  "host": "report.example.com",
  "proxy": "localhost",
  "reported_at": "2021-10-08T07:15:49.341191",
  "results": [
    {
      "result": {
        "msg": "Nothing to do",
        "changed": false,
        "rc": 0,
        "invocation": {
          "module_args": {
            "name": [
              "vim"
            ],
            "state": "present",
            "allow_downgrade": false,
            "autoremove": false,
            "bugfix": false,
            "disable_gpg_check": false,
            "download_only": false,
            "installroot": "/",
            "install_repoquery": true,
            "install_weak_deps": true,
            "security": false,
            "skip_broken": false,
            "update_cache": false,
            "update_only": false,
            "validate_certs": true,
            "lock_timeout": 30
          }
        },
        "_ansible_no_log": false
      },
      "task": {
        "action": "ansible.builtin.package",
        "async_val": 0,
        "async": 0,
        "delay": 5,
        "poll": 15,
        "retries": 3,
        "name": "",
        "connection": "smart",
        "check_mode": false,
        "diff": false,
        "any_errors_fatal": false,
        "throttle": 0,
        "become": false,
        "become_method": "sudo",
        "uuid": "1c697a0b-ae03-a393-94fb-000000000009",
        "finalized": true,
        "squashed": true
      },
      "failed": false,
      "level": "info",
      "friendly_message": "Package(s) vim are present: Nothing to do"
    }
  ],
  "keywords": [

  ],
  "telemetry": {
    "parse": 0.03382004797458649,
    "process": 0.015477067790925503
  },
  "status": {
    "applied": 0,
    "failed": 0,
    "skipped": 0
  },
  "check_mode": false
}
```

ofedoren replied:
Yeah, exactly: you have here `["task"]["action"] == "ansible.builtin.package"` the action is the same you've used to run ansible `-m ansible.builtin.package`. But when you run a playbook you will have `["task"]["action"] == "package"`. See last line:
![ScreenShot-1633693643852](https://user-images.githubusercontent.com/32508194/136551633-58a3b9d4-08e1-4382-9516-fbbf9b3de838.png)

ofedoren replied:
This and all below is still not resolved...

UPD: Actually we could support both `ansible.builtin.module` and just `module`:

```ruby
# ...
case task_tree["action"].split('.').last
when "package"
# ...
# ...
```

lzap replied:
No, I do not have `"action": "package"`, see here and in other example dumps:

https://github.com/theforeman/smart_proxy_host_reports/pull/5/files#diff-d22ea879447ac1084da738dca64815331e0ae8b1d72e4ef8151144b59f4a7c8bR51

Maybe you had an old version of the callback?

I am not sure about splitting it by dot, what is added benefit? Looks like Ansible uses some kind of dot-syntax for module names, so let's just use it.

ofedoren commented on lib/smart_proxy_host_reports/ansible_processor.rb:
```diff
@@ -82,46 +74,56 @@ def spool_report
+    when "ansible.builtin.user"
+      msg = "User #{result_tree["name"]} is #{result_tree["state"]} with uid: #{result_tree["uid"]}"
+    when "ansible.builtin.cron"
+      msg = "Cron job: #{module_args_tree["minute"]} #{module_args_tree["hour"]} #{module_args_tree["day"]} #{module_args_tree["month"]} #{module_args_tree["weekday"]} #{module_args_tree["job"]} (disabled: #{module_args_tree["disabled"]})"
```
```suggestion
      msg = "Cron job: #{module_args_tree["minute"]} #{module_args_tree["hour"]} #{module_args_tree["day"]} #{module_args_tree["month"]} #{module_args_tree["weekday"]} #{module_args_tree["job"]} and disabled: #{module_args_tree["disabled"]}"
```
If you're changing the way we had, let's stick with the same approach :)

lzap replied:
> If you're changing the way we had, let's stick with the same approach :)

Why? If we want to change anything, this is the right time. This is a complete rewrite, format is new, UI will be new. I am not following. I can do that, just explain the reason.

ofedoren replied:
Yes, that's OK. What I meant by that is that above you have no parentheses, but in this sentece you have them. My comment and suggestion is about to have consintency in the new way of friendly messages.

lzap replied:
Oh sure ok fixing.

ofedoren reviewed (commented):
LGTM, but I have few suggestions.

lzap commented:
Amended your suggestion and also created this little baby:

```
[lzap@nuc smart_proxy_host_reports]$ contrib/upload-fixture
Usage:
  contrib/upload-fixture -h               Display this help message
  contrib/upload-fixture -u URL           Proxy URL (http://localhost:8448)
  contrib/upload-fixture -f FILE          Fixture to upload
  contrib/upload-fixture -a               Upload all fixtures
[lzap@nuc smart_proxy_host_reports]$ contrib/upload-fixture -a
contrib/fixtures/ansible-copy-nochange.json: 200
contrib/fixtures/ansible-copy-success.json: 200
contrib/fixtures/ansible-package-install-failure.json: 200
contrib/fixtures/ansible-package-install-nochange.json: 200
contrib/fixtures/ansible-package-install-success.json: 200
contrib/fixtures/ansible-package-remove-failure.json: 200
contrib/fixtures/ansible-package-remove-success.json: 200
```

There is now a new development option to capture fixtures so we can create them. See the README for futher details.

However, for some reason Puppet does not work - YAML is sent without newlines, therefore the `data.gsub(YAML_CLEAN, "")` resolves to just `---` (deletes everyting to the end of the line) and this ends up with an error. I do not understand why this is happening, maybe you can take a look?

ofedoren commented on lib/smart_proxy_host_reports/host_reports_api.rb:
```diff
@@ -31,6 +31,10 @@ def check_content_type(format)
       check_content_type(format)
       input = request.body.read
+      if Proxy::HostReports::Plugin.settings.incoming_save_dir
+        filename = File.join(Proxy::HostReports::Plugin.settings.incoming_save_dir, "#{format}-#{SecureRandom.uuid}")
```
Not a request, but more like suggestion: maybe instead of UUID we could use a timestamp?

lzap replied:
Good point, sure.

ofedoren commented on .gitignore:
```diff
@@ -1 +1,2 @@
 smart_proxy_host_reports-*.gem
+/contrib/upload-fixture.conf
```
If your script supports config files, why we hide it? I guess it won't harm to have something like `/contrib/upload-fixture.conf.example` file with defaults that can be adjusted and renamed to `/contrib/upload-fixture.conf`.

lzap replied:
Yeah let's add an example. I thought it will be rather rare for others to use it - I do not have HTTP port set to 8000 due to historical reasons.

ofedoren commented on settings.d/host_reports.yml.example:
```diff
@@ -13,4 +13,10 @@
+
+# Development settings (do not use)
+# Override hostnames of incoming reports
+#:override_hostname: report.example.com
```
If I don't touch it as a user (not a developer) then I will encounter:
```
2021-10-08T15:36:42  [W] Error details for Disabling all modules in the group ['host_reports'] due to a failure in one of them: Parameter 'override_hostname' is expected to have a non-empty 
value: <Proxy::Error::ConfigurationError>: Parameter 'override_hostname' is expected to have a non-empty value
/home/vagrant/smart-proxy/lib/proxy/plugin_validators.rb:41:in `validate!'
/home/vagrant/smart-proxy/lib/proxy/plugin_validators.rb:21:in `evaluate_predicate_and_validate!'
/home/vagrant/smart-proxy/lib/proxy/plugin_initializer.rb:351:in `block in execute_validators'
/home/vagrant/smart-proxy/lib/proxy/plugin_initializer.rb:348:in `each'
/home/vagrant/smart-proxy/lib/proxy/plugin_initializer.rb:348:in `inject'
/home/vagrant/smart-proxy/lib/proxy/plugin_initializer.rb:348:in `execute_validators'
/home/vagrant/smart-proxy/lib/proxy/plugin_initializer.rb:341:in `validate_settings'
/home/vagrant/smart-proxy/lib/proxy/plugin_initializer.rb:288:in `load_settings'
/home/vagrant/smart-proxy/lib/proxy/plugin_initializer.rb:263:in `load_plugin_settings'
/home/vagrant/smart-proxy/lib/proxy/plugin_initializer.rb:80:in `load_plugin_settings'
/home/vagrant/smart-proxy/lib/proxy/plugin_initializer.rb:167:in `each'
/home/vagrant/smart-proxy/lib/proxy/plugin_initializer.rb:167:in `initialize_plugins'
/home/vagrant/smart-proxy/lib/launcher.rb:171:in `launch'
/home/vagrant/smart-proxy/bin/smart-proxy:6:in `<top (required)>'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `load'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `kernel_load'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:28:in `run'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:476:in `exec'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:30:in `dispatch'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:24:in `start'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/libexec/bundle:46:in `block in <top (required)>'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/lib/bundler/friendly_errors.rb:123:in `with_friendly_errors'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/bundler-2.1.4/libexec/bundle:34:in `<top (required)>'
/opt/rh/rh-ruby27/root/usr/bin/bundle:23:in `load'
/opt/rh/rh-ruby27/root/usr/bin/bundle:23:in `<main>'
```

lzap replied:
Fixed.

ofedoren reviewed (commented):
Thanks, @lzap! I appreciate new script and fixtures for test purposes, it surely can help when there is no time to set up a new env :)

But I still see few issues though...

lzap commented:
I rebased your remarks, also upload script now also recognizes single files.

ofedoren commented:
@lzap, if I use
```sh
ansible-playbook /home/vagrant/foreman_ansible_test/test.yml
```
then I have this:
![ScreenShot-1634042559178](https://user-images.githubusercontent.com/32508194/136958588-5d05c255-26e4-4c4f-852f-269ee91fdcdf.png)

and when I use
```sh
ansible sppup.nuka.com -m ansible.builtin.copy -a "src=/etc/hosts dest=/tmp/hosts"
```
I have this:
![ScreenShot-1634042735676](https://user-images.githubusercontent.com/32508194/136958923-e81c51db-9f71-4f7e-9cec-51c8262d49ef.png)


Please, look at the `copy` tasks:
 - In the first example (playbook) I see task name, but the friendly description is missing
 - In the second example (using builtin stuff via direct ansible) I don't see task name (which is OK), but I see friendly description.
 
 UPD: This was tested with latest devel branch of `foreman-ansible-modules` and `foreman_host_reports`, and this patch as is now.

lzap commented:
I am assuming you refer to https://github.com/ares/foreman_ansible_test/blob/main/tasks/main.yaml

Yeah I did not test against full Ansible play yet. Let me do that and also capture this for testing purposes too.

lzap commented:
Rebased, fixed the problem. I could use regexp but these are slow to compile, so I am going for a very quick string case.

Added also "harmless tests", for now only Ansible one. No need for sudo or anything - it is very quick as well. Also gathers facts which I am dropping for now and leaving a TODO item in the code. Also captured harmless test runs so they can be uploaded easily too.

Please ignore the friendly messages, it is just a prototype, Anna will be working on these soon. The code is ugly it needs to be refactored.

ofedoren commented on lib/smart_proxy_host_reports/ansible_processor.rb:
```diff
@@ -82,46 +75,61 @@ def spool_report
-    return log["msg"] if log["failed"]
+  def process_message(result)
+    msg = "N/A"
+    return msg if result["task"].nil? || result["task"]["action"].nil?
+    return result["result"]["msg"] if result["failed"]
```
We don't store the returned value anywhere. Maybe it's better to do like this?
```suggestion
    return result["friendly_message"] = msg if result["task"].nil? || result["task"]["action"].nil?
    return result["friendly_message"] = result["result"]["msg"] if result["failed"]
```

ofedoren reviewed (commented):
Thanks, @lzap! LGTM and works as expected. I have the last concern, I promise :)

lzap commented:
Sure makes sense, thanks!

ofedoren reviewed (approved):
Thanks! No other concerns. Let's move it :)

lzap commented:
Smash that button, do you have permissions? If not that is an overlook!

lzap commented:
Can we get push rights to Oleh, @melcorr ? Thank you!

melcorr commented:
Hope that's OK!

ofedoren commented:
Thanks, @lzap, merging :) If anything will occure, we will fix it in a smaller PR.

Thanks, @melcorr, for the rights :)

avitova commented on contrib/upload-fixture:
```diff
@@ -0,0 +1,66 @@
+do_curl() {
+  curl -s -H "Content-Type: $1" \
+    -X POST $URL/host_reports/$2 \
+    -d @$3 -w " - $3: %{http_code}\n"
```
Looks like this should work with the --data-binary option instead of the "-d". According to curl manual, "-d" does not preserve new-lines.
